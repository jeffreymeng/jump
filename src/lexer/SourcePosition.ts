/**
 * A SourcePosition refers to a specific position in the source.
 * Each character in the source has a unique position.
 */
export default class SourcePosition {
	/**
	 * @param source - The text of the source code.
	 * @param index - The index of the character within the source text.
	 */
	constructor(protected readonly source: string, protected readonly index: number) {
		if (index >= source.length) {
			throw new Error("SourcePosition expected an index less than the length of source.")
		}
	}

	/**
	 * Returns the line of the index in the given source text. Lines are 1-indexed.
	 */
	public getLine(): number {
		return this.source.substring(0, this.index).split("\n").length
	}

	/**
	 * Returns the index of the character within its line. Columns are
	 * 1-indexed.
	 */
	public getCol(): number {
		// given the column, look backwards for the closest newline strictly
		// before the current index, since if the current character is a
		// newline that doesn't count.
		return this.index - this.source.lastIndexOf("\n", this.index - 1);
	}

}
