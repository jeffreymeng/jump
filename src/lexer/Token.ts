import SourcePosition from "./SourcePosition";

export enum TokenType {
	STRING_LITERAL = "string_literal",
	INT_LITERAL = "int_literal",
	DOUBLE_LITERAL = "double_literal",
	OPERATOR = "operator",
	IDENTIFIER = "identifier",
	CONTROL = "control",
	KEYWORD = "keyword",
}
export default class Token {
	/**
	 *
	 * @param type - The type of the token
	 * @param symbol - The value of the token
	 */
	constructor(
		public readonly type: TokenType,
		public readonly symbol: string,
		public readonly position: SourcePosition
	) {}

	/**
	 * Returns whether or not the token matches another token, in terms of content,
	 * but not position.
	 * @param other
	 */
	public equals(other: Token) {
		return this.type === other.type && this.symbol === other.symbol;
	}

	/**
	 * Returns whether or not the token matches the provided type and, if provided, symbol.
	 * @param type
	 * @param symbol
	 */
	public is(type: TokenType, symbol?: string) {
		return this.type === type && (!symbol || this.symbol === symbol);
	}

	public toString() {
		return `Token<${this.type}, '${this.symbol}'>`;
	}
}
