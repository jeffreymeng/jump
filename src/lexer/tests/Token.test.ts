import { describe, expect, test } from "@jest/globals";
import Token, { TokenType } from "../Token";
import SourcePosition from "../SourcePosition";

describe("Token", () => {
	const position = new SourcePosition("", 0);
	test("creating new token", () => {
		const t = new Token(TokenType.KEYWORD, "for", position);
		expect(t.type).toBe(TokenType.KEYWORD);
		expect(t.symbol).toBe("for");
	});

	test("equality", () => {
		const t = new Token(TokenType.KEYWORD, "for", position);
		const t2 = new Token(TokenType.CONTROL, "for", position);
		const t3 = new Token(TokenType.KEYWORD, "notfor", position);
		const t4 = new Token(TokenType.STRING_LITERAL, "adsf", position);
		const t5 = new Token(TokenType.KEYWORD, "for", position);

		expect(t.equals(t2)).toBe(false);
		expect(t.equals(t3)).toBe(false);
		expect(t.equals(t4)).toBe(false);
		expect(t.equals(t5)).toBe(true);
	});

	test("is", () => {
		const t = new Token(TokenType.KEYWORD, "for", position);

		expect(t.is(TokenType.KEYWORD, "for")).toBe(true);
		expect(t.is(TokenType.CONTROL, "for")).toBe(false);
		expect(t.is(TokenType.KEYWORD, "forr")).toBe(false);
	});

	test("toString", () => {
		const op = new Token(TokenType.OPERATOR, "+", position);
		const str = new Token(TokenType.STRING_LITERAL, "foo", position);
		const int = new Token(TokenType.INT_LITERAL, "5", position);

		expect(op.toString()).toMatchSnapshot();
		expect(int.toString()).toMatchSnapshot();
		expect(int.toString()).toMatchSnapshot();
	});
});
