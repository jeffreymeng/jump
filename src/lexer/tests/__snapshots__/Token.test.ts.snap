// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token toString 1`] = `"Token<operator, '+'>"`;

exports[`Token toString 2`] = `"Token<int_literal, '5'>"`;

exports[`Token toString 3`] = `"Token<int_literal, '5'>"`;
